compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/TransformUtils.kt

```diff
@@ -96,13 +101,24 @@ internal fun FirValueParameter.transformVarargTypeToArrayType() {
 }
 
 internal fun FirTypedDeclaration.transformTypeToArrayType() {
+    val returnTypeRef = this.returnTypeRef
+    require(returnTypeRef is FirResolvedTypeRef)
+    // If the delegated type is already resolved, it means we have already created a resolved array type for this vararg type declaration.
+    // This is because in the buildResolvedTypeRef call below, we set the delegated type ref to the previous (non-vararg) resolved type ref.
+    if (returnTypeRef.delegatedTypeRef is FirResolvedTypeRef) return
```

Comment:
To be safer (and match with comments), I think it's good to double-check if `delegatedTypeRef`'s `source` is `ArrayTypeFromVarargParameter`.